lib/web_catalog/api_extractor.es6.js

```diff
@@ -0,0 +1,406 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+foam.CLASS({
+  name: 'apiExtractor',
+  package: 'com.web.catalog',
+  description: 'Extract api catalog from a object graph',
+  properties: [
+    {
+      name: 'defaults',
+      description: 'Default value of configuration',
+      factory: function() {
+        return {
+          // When reatainBuildInMembers is true,
+          // build-in properties of function and object
+          // are returned as APIs.
+          retainBuildInMembers: false,
+          // When retainConstantMembers is false,
+          // constan primitive properties are filtered.
+          retainConstantMembers: false,
+        };
+      },
+    },
+    {
+      name: 'blacklistProperties',
+      description: 'Blacklisted properties will not visited or stored',
+      factory: function() {
+        return [
+          // toString is a default method of any objects.
+          'toString',
+          // Constructors have prototype but are not interfaces.
+          'constructor',
+          // These are known bugs in Firefox.
+          'CSS2Properties',
+        ];
+      },
+    },
+    {
+      name: 'constantTypes',
+      description: 'Properties of these types with writable 0 are filtered.' +
+        'if defaults.retainConstantMembers is false',
+      factory: function() {
+        return ['boolean', 'number', 'string'];
+      },
+    },
+    {
+      name: 'buildInFunctionProperties',
+      description: 'Properties of Function.prototype are considered' +
+        'as boring properties which exists in all function objects.',
+      factory: function() {
+        return ['prototype'];
+      },
+    },
+    {
+      name: 'buildInObjectProperties',
+      description: 'Properties of Object.prototype are considered' +
+        'as boring properties which may exists in all objects.',
+      factory: function() {
+        return [];
+      },
+    },
+    {
+      name: 'metadata',
+      description: 'Contains infomation of Object, Function and roots ids.',
+      factory: function() {
+        return {
+          rootId: null,
+          objectProtoId: null,
+          functionProtoId: null,
+          objectId: null,
+          functionId: null,
+        };
+      },
+    }
+  ],
+  methods: [
+    {
+      name: 'getObjectProperties_',
+      description: 'A wrraper function to get properties with constant,' +
+        'non-own-properties filtered.',
+      code: function(og, id) {
+        let keys = og.getObjectKeys(id);
+        keys = keys.filter((key) => {
+          let keyId = og.lookup(key, id);
+          let meta = og.lookupMetaData(key, id);
+          if (Object.keys(meta).length === 0) {
+            // If meta contains no information, it means this property
+            // is not own property but inherited from its prototype.
```

Comment:
When you say "object-graph is modified", what do you mean? Are you using a non-master branch of `object-graph-js`? Is there code somewhere that's changing the input object graph before it's passed to this component?